tests/assoc_sim.R

```diff
@@ -0,0 +1,158 @@
+                help = "Filename for RAPPOR parameters"),
+    make_option(c("--reports", "-r"), default = "reports.csv",
+                help = "Filename for reports"),
+    make_option(c("--map", "-m"), default = "map",
+                help = "Filename *prefix* for map(s)"),
+    make_option(c("--num", "-n"), default = 1e05,
+                help = "Number of reports"),
+    make_option(c("--unif", "-u"), default = FALSE,
+                help = "Run simulation with uniform distribution")
+  )
+  opts <- parse_args(OptionParser(option_list = option_list))
+}    
+
+source("../analysis/R/encode.R")
+source("../analysis/R/decode.R")
+source("../analysis/R/simulation.R")
+source("../analysis/R/read_input.R")
+source("../analysis/R/association.R")
+
+# Read candidates from a csv file
+# Inputs: filename. The file is expected to contain two rows of strings
+#         (one for each variable):
+#         "google.com", "apple.com", ...
+#         "ssl", "nossl", ...
+# Returns: a list containing strings
+GetCandidatesFromFile <- function(filename) {
+  contents <- read.csv(filename, header = FALSE)
+  # Expect 2 rows of candidates
+  if(nrow(contents) != 2) {
+    stop(paste("Candidate file", filename, "expected to have
+               two rows of strings."))
+  }
+  list(var1 = as.vector(t(contents[1,])),
+       var2 = as.vector(t(contents[2,])))
+}
+
+# Simulate correlated reports and write into reportsfile
+# Inputs: N = number of reports
+#         candidates = list containing a list of candidate strings
+#         params = list with RAPPOR parameters
+#         unif = whether to replace poisson with uniform
+#         mapfile = file to write maps into (with .csv suffixes)
+#         reportsfile = file to write reports into (with .csv suffix)
+SimulateReports <- function(N, candidates, params, unif,
+                            mapfile, reportsfile) {
+  # Compute true distribution
+  m <- params$m  
+  samples <- list()
+
+  if (unif) {
+    # Draw uniformly from 1 to 10
+    samples[[1]] <- as.integer(runif(N, 1, 10))
+  } else {
+    # Draw from a Poisson random variable
+    samples[[1]] <- rpois(N, 1) + rep(1, N)
+  }
+  
+  # Pr[var2 = N + 1 | var1 = N] = 0.5
+  # Pr[var2 = N     | var1 = N] = 0.5
+  samples[[2]] <- samples[[1]] + sample(c(0, 1), N, replace = TRUE)
+  
+  # Replace sample i with candidate i for each variable
+  for(i in 1:2) {
+    if(max(samples[[i]]) <= length(candidates[[i]])) {
+      samples[[i]] <- candidates[[i]][samples[[i]]]
+    } else {
+      # Pad candidates with sample strings
+      # If only 2 candidates, new set of candidates becomes:
+      # candidate1, candidate2, str3, str4, ...
+      len <- length(candidates)
+      candidates[[i]][(len + 1):max(samples[[i]])] <- apply(
+        as.matrix((len + 1):max(samples[[i]])),
+        1,
+        function(i) paste("str", as.character(i), sep = ""))
+      samples[[i]] <- candidates[[i]][samples[[i]]]
```

Comment:
Please write this in a more functional rather than imperative style to follow the rest of the code.  You are first assigning samples[[1]], then samples[[2]] as a function of samples[[1]], then overwriting both of them again.  A functional style is more suited to mathematics, and it makes it so each variable is read-only, and has a meaningful variable name, rather than something like samples[[1]] and samples[[2]].

After you've built them, you can do:

samples <- list(foo, bar)